host/sgx/sgxload.c

```diff
@@ -355,18 +359,53 @@ oe_result_t oe_sgx_create_enclave(
         if (oe_sgx_is_simulation_load_context(context))
         {
             /* Allocation memory-mapped region */
-            if (!(base = _allocate_enclave_memory(enclave_size)))
+            if (!(image_base = _allocate_enclave_memory(enclave_size)))
                 OE_RAISE(OE_OUT_OF_MEMORY);
         }
 #else
         // Wrong code path
         result = OE_UNSUPPORTED;
         goto done;
-#endif // OEHOSTMR
+#endif // !defined(OEHOSTMR)
     }
 
+    /*
+     * Load desired enclave start address. NOTE: Currently, this value is NULL
+     * when zero base enclave is not enabled. Also, start_address has to be
+     * aligned to OE_PAGE_SIZE.
+     */
+    start_address = (void*)context->start_address;
+    if ((uint64_t)start_address % OE_PAGE_SIZE)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (context->create_zero_base_enclave)
+    {
+        /*
+         * enclave_image_address is the base address of the image to be
+         * loaded into enclave.
+         */
+        enclave_elrange.enclave_image_address = (uint64_t)start_address;
+        /*
+         * elrange_start_address is base address of the enclave address
+         * range, which needs to be lower than enclave_image_address.
+         */
+        enclave_elrange.elrange_start_address = (uint64_t)OE_ADDRESS_ZERO;
+        /*
+         * elrange_size is the size of the enclave address range.
+         */
+        enclave_elrange.elrange_size =
+            oe_round_u64_to_pow2(enclave_size + (uint64_t)start_address);
```

Comment:
The non-zero-base case does not round up enclave_size to power of 2 on line 404. Is it necessary to round up here for zero-base case?

Reply:
It was enforced by existing logic:
https://github.com/openenclave/openenclave/blob/b7ef81ef772bc3d0ebfaa47c596b887cb16604c8/host/sgx/create.c#L368

Reply:
In that case, `elrange_size` for zero-base enclave might be set unnecessarily large. Say `*loaded_enclave_pages_size`= 129MB, for the zero-base enclave, if the enclave_start_addr is 100MB, then `elrange_size`=512MB with the current implementation. If we skip line 368 in create.c, `elrange_size` would be 256MB.

Reply:
or we should do
`oe_round_u64_to_pow2(enclave_commit_size+ (uint64_t)start_address);`
instead, using the unrounded value.

Reply:
@RRathna please check

Reply:
Yes, using enclave_commit_size solves the problem of allocating larger than required range. Changes incorporated.